### PR TITLE
Adjust detection of UIKit.UIView.Dispose invocations

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the current behavior?

Use code calling UIView.Dispose can break the runtime with double dispose issues like https://github.com/xamarin/xamarin-macios/issues/19493.

## What is the new behavior?

Uno now provides UNO0002 which will indicate when user code is invoking `UIView.Dispose()` and `UIView.Dispose(bool)` in order to detect incorrect uses early.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 4df839e</samp>

This pull request introduces a new analyzer `UnoDoNotDisposeNativeViews` to warn against disposing native views on iOS, and deprecates the `NSObjectExtensions.ValidateDispose` method, which was causing issues. It also updates the documentation and the Roslyn dependencies for the analyzers.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by jeanplevesque — APPROVED** [review]

**Comment by jeromelaban** [comment]
@mergifyio backport release/stable/5.0 release/stable/4.10 legacy/4x

File: src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+			description: Description,
+			helpLinkUri: "https://aka.platform.uno/UNO0002"
+		);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```
Comment (Youssef1313) [inline src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs:41]: Do we really want to report the diagnostic in generated code?

File: src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+					&& (
+						invocationOperation.TargetMethod.Parameters.Length == 0
+						|| (
+							invocationOperation.TargetMethod.Parameters.Length == 1
+							&& invocationOperation.TargetMethod.Parameters[0].Type.SpecialType == SpecialType.System_Boolean)))
+				{
+					var diagnostic = Diagnostic.Create(
+						Rule,
+						context.Operation.Syntax.GetLocation(),
+						invocationOperation.TargetMethod.ToDisplayString()
```
Comment (Youssef1313) [inline src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs:75]: Would the error message say "Don't call Dispose() on UIKit.UIView.Dispose()"? That sounds a bit strange.

File: src/SourceGenerators/Uno.UI.SourceGenerators/DependencyObject/DependencyObjectGenerator.cs [hunk 3]
```diff
@@ -569,9 +569,21 @@ private void WriteDispose(INamedTypeSymbol typeSymbol, IndentedStringBuilder bui
...
+						// 
+						// This operation can fail if Dispose() is called by user code on UIView instances.
+						// The Roslyn analyzer UnoDoNotDisposeNativeViews will warn the developer if this is the case.
+						// 
+						// For native exceptions that may be raised if Disposed is called incorrectly, see 
+						// https://github.com/xamarin/xamarin-macios/issues/19493.
+
 						if(_isDisposed)
 						{{
 							base.Dispose(disposing);
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/DependencyObject/DependencyObjectGenerator.cs:589]: Will this call produce a diagnostic? In this case we might want to skip analyzing generated code.

**Review by Youssef1313 — APPROVED** [review]

**Review by ajpinedam — APPROVED** [review]

File: src/SourceGenerators/Uno.UI.SourceGenerators/DependencyObject/DependencyObjectGenerator.cs (hunk 3 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/DependencyObject/DependencyObjectGenerator.cs:589]: Indeed, and there are other legitimate scenarios where this is useful, so technically, base.Dispose, when called from Dispose it valid.

File: src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs (hunk 2 above)
Comment (jeromelaban) [reply src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs:75]: True, will adjust!

File: src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs (hunk 1 above)
Comment (jeromelaban) [reply src/Uno.Analyzers/UnoDoNotDisposeNativeViews.cs:41]: I'm not sure. There may be generators that create that pattern...

File: src/Uno.Analyzers.Tests/UnoDoNotDisposeNativeViewsTests.cs [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+			"""
+			namespace UIKit
+			{
+					public class UIView : IDisposable
+					{ 
+						public void Dispose() { }
+						public void Dispose(bool disposing) { }
+					}
+			}
+			""";
```
[the remark this answers is not in the document]
Comment (jeromelaban) [reply src/Uno.Analyzers.Tests/UnoDoNotDisposeNativeViewsTests.cs:24]: I don't think that's needed for this particular test. The analyzer does not rely on the structure of `UIView` other than `Dispose()` and `Dispose(bool)`. We'll adjust as needed.

**Pull request merged**